### PR TITLE
improve the `GOV` generator to generate more interesting signals

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
# Description

Improve the GOV signal generator to make sure we generate signals that can depend on any part of the proposals tree.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/V2/Conway/GOV.hs [hunk 1]
```diff
@@ -283,51 +276,47 @@ govProceduresSpec ge@GovEnv {..} ps =
...
     govAction
     -- ParameterChange
     ( branch $ \mPrevActionId ppUpdate policy ->
-        -- TODO: this should be any ppup proposal in the tree
-        [ assert $ mPrevActionId ==. lit (gePrevGovActionIds ^. grPParamUpdateL)
+        [ assert $ mPrevActionId `elem_` lit ppupIds
         , wfPParamsUpdate ppUpdate
         , assert $ policy ==. lit gePPolicy
         ]
     )
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/V2/Conway/GOV.hs]: I am not a big fan of using field puns and then using Foo{..}, there is no way to tell where the fields are coming from (or even if they are a field. Can't we say something like Foo{ ppupIds }, so then we know it is coming from Foo?.
Every symbol should have a visible binding site. If there is no way to do this, then we could resort to
t@Foo{..}      then use (ppupIds t)

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/V2/Conway/GOV.hs [hunk 2]
```diff
@@ -283,51 +276,47 @@ govProceduresSpec ge@GovEnv {..} ps =
...
+              reify mPrevActionId hfIdMinorVer $ \currentMinorVersion ->
+                caseOn
+                  mSuccMajorVersion
+                  ( branch $ \_ ->
+                      [ majorVersion ==. currentMajorVersion
+                      , minorVersion ==. currentMinorVersion + 1
+                      ]
+                  )
+                  ( branch $ \majorVersion' ->
+                      [ assert $ majorVersion `member_` (singleton_ currentMajorVersion <> singleton_ majorVersion')
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/V2/Conway/GOV.hs:314]: I guess we really would like some kind of 'or' here 
majorVersion ==. currentMajorVersion ||. majorVersion ==. majorVersion'

File: libs/constrained-generators/src/Constrained/Spec/Generics.hs [hunk 3]
```diff
@@ -383,6 +384,24 @@ constrained' ::
...
+  , IsProd (SimpleRep b)
+  , HasSpec fn a
+  , HasSpec fn b
+  , IsPred p fn
+  ) =>
+  Term fn a ->
+  (a -> b) ->
+  FunTy (MapList (Term fn) (Args (SimpleRep b))) p ->
+  Pred fn
+reify' a r f = reify a r $ \x -> match @fn @p x f
```
Comment (TimSheard) [inline libs/constrained-generators/src/Constrained/Spec/Generics.hs:406]: Maybe a comment about what the difference is between  reify and reify'

**Review by TimSheard — APPROVED** [review]
Every things looks reasonable.  Please, if at all possible, make binding sites visible.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/V2/Conway/GOV.hs (hunk 2 above)
Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/V2/Conway/GOV.hs:314]: yeah, that or `list_ :: [Term fn a] -> Term fn [a]`

Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/V2/Conway/GOV.hs:314]: Which can be built if we add `cons_ :: Term fn a -> Term fn [a] -> Term fn [a]`

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/V2/Conway/GOV.hs (hunk 1 above)
Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/V2/Conway/GOV.hs]: `ppupIds` is bound in a `where` clause at the bottom of the spec.

**Pull request merged**